ddl/ddl_tiflash_api.go

```diff
@@ -0,0 +1,453 @@
+	}
+	// Try to update.
+	for _, kv := range resp.Kvs {
+		if string(kv.Key) == key {
+			origin = string(kv.Value)
+			break
+		}
+	}
+	if origin != httpAddr {
+		logutil.BgLogger().Warn(fmt.Sprintf("Update status addr of %v from %v to %v", key, origin, httpAddr))
+		err := ddlutil.PutKVToEtcd(d.ctx, d.etcdCli, 1, key, httpAddr)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return nil
+}
+
+func updateTiFlashStores(pollTiFlashContext *TiFlashManagementContext) error {
+	// We need the up-to-date information about TiFlash stores.
+	// Since TiFlash Replica synchronize may happen immediately after new TiFlash stores are added.
+	tikvStats, err := infosync.GetTiFlashStoresStat(context.Background())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	pollTiFlashContext.TiFlashStores = make(map[int64]helper.StoreStat)
+	for _, store := range tikvStats.Stores {
+		for _, l := range store.Store.Labels {
+			if l.Key == "engine" && l.Value == "tiflash" {
+				pollTiFlashContext.TiFlashStores[store.Store.ID] = store
+				logutil.BgLogger().Debug("Found tiflash store", zap.Int64("id", store.Store.ID), zap.String("Address", store.Store.Address), zap.String("StatusAddress", store.Store.StatusAddress))
+			}
+		}
+	}
+	logutil.BgLogger().Info("updateTiFlashStores finished", zap.Int("TiFlash store count", len(pollTiFlashContext.TiFlashStores)))
+	return nil
+}
+
+func (d *ddl) pollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContext *TiFlashManagementContext) (bool, error) {
+	allReplicaReady := true
+	defer func() {
+		pollTiFlashContext.HandlePdCounter += 1
+		pollTiFlashContext.HandlePdCounter %= PullTiFlashPdTick
+	}()
+
+	updateTiFlash := pollTiFlashContext.UpdateTiFlashStoreCounter%UpdateTiFlashStoreTick == 0
+	if updateTiFlash {
+		if err := updateTiFlashStores(pollTiFlashContext); err != nil {
+			// If we failed to get from pd, retry everytime.
+			pollTiFlashContext.UpdateTiFlashStoreCounter = 0
+			return false, errors.Trace(err)
+		}
+	}
+	pollTiFlashContext.UpdateTiFlashStoreCounter += 1
+	pollTiFlashContext.UpdateTiFlashStoreCounter %= UpdateTiFlashStoreTick
+
+	// The following loop updates TiFlash store's status address.
+	for _, store := range pollTiFlashContext.TiFlashStores {
+		s := store
+		if err := d.UpdateTiFlashHTTPAddress(&s); err != nil {
+			logutil.BgLogger().Error("Update TiFlash status address failed", zap.Error(err))
+		}
+	}
+
+	// Start to process every table.
+	schema := d.GetInfoSchemaWithInterceptor(ctx)
+	if schema == nil {
+		return false, errors.New("Schema is nil")
+	}
+
+	var tableList = make([]TiFlashReplicaStatus, 0)
+
+	// Collect TiFlash Replica info, for every table.
+	for _, db := range schema.AllSchemas() {
+		tbls := schema.SchemaTables(db.Name)
+		for _, tbl := range tbls {
+			tblInfo := tbl.Meta()
+			GetTiFlashReplicaInfo(tblInfo, &tableList)
+		}
+	}
+
+	// Sync pd rules.
+	handlePd := pollTiFlashContext.HandlePdCounter%PullTiFlashPdTick == 0
+	if handlePd {
```

Comment:
This loop is aimed to delete useless pd rules and add missing pd rules. However the dependent function `getDropOrTruncateTableTiflash` is broken(it can't handle partitioned tables). Should we [fix this function(quite ugly and maybe buggy)](https://github.com/CalvinNeo/tidb/tree/part2.tryfixpdrules), or just delete the loop? 

My worries about deleting the loop, is when there are failures when gc.

Reply:
Maybe we can first delete the loop and add back the correct one if need?